### PR TITLE
Input HAL

### PR DESCRIPTION
**Pull request by Victor-C-Zhang**
Draft the input HAL. Implement a cin-based input "device" for Linux. Integrate it with the base service.

### CODE REVIEW

File: device/common/input/hal/input.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#pragma once
+
+#include <cstdint>
+#include <string>
+
+namespace hal {
+namespace input {
+
+enum input_err_t : uint16_t {
+  RES_NOOP = 1000,
```
Comment (q349wang) [inline device/common/input/hal/input.h:12]: Why 1000?

File: device/linux/input/input.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+static input_callbacks* cb = nullptr;
+std::thread* input_thr;
+
+void input_loop() {
+  std::string s;
+  while (1) {
+    std::cin >> s;
+    std::cout << "Got input \"" << s << "\"" << std::endl;
+    if (s == "ls") {
+      cb->on_left_short_press();
```
Comment (q349wang) [inline device/linux/input/input.cpp:20]: Should we check these are set?

File: device/linux/input/input.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+input_err_t input_init(input_callbacks* callbacks) {
+  cb = new input_callbacks(*callbacks);
+  input_thr = new std::thread(input_loop);
+  input_thr->detach();
+  return RES_OK;
+}
+
+input_err_t input_destruct() {
+  delete cb;
```
Comment (q349wang) [inline device/linux/input/input.cpp:48]: Shouldn't we delete the CB only after the input loop stops or are we ok with the potential default here

**Review by q349wang — APPROVED** [review]
Some comments, lgtm

File: device/common/input/hal/input.h (hunk 1 above)
Comment (Victor-C-Zhang) [reply device/common/input/hal/input.h:12]: big number

File: device/linux/input/input.cpp (hunk 3 above)
Comment (Victor-C-Zhang) [reply device/linux/input/input.cpp:48]: this impl is missing a lot of pieces. this is really just a demo to make sure it sort of works. let me think a bit more about this later if we want to productionize it

**Pull request merged**